File: regression/cs/tst5.cs
using Microsoft.Z3;
using System;

class Test
{

    public void Run()
    {        
        using (Context ctx = new Context())
        {
            FuncDecl f = ctx.MkFuncDecl("f", new Sort[] { ctx.IntSort, ctx.RealSort }, ctx.IntSort);

            try
            {
                Console.WriteLine(f.Domain[3]);
            }
            catch (IndexOutOfRangeException ex)
            {
                Console.WriteLine("failed: " + ex.Message);
            }

            IntExpr x = ctx.MkIntConst("x");

            Console.WriteLine(f[ctx.MkInt(1), ctx.MkReal(1)]);
            Console.WriteLine(f[ctx.MkInt(1), ctx.MkReal(1)].Sort);
            Console.WriteLine(f[ctx.MkInt(1), ctx.MkReal(1)].NumArgs);
            foreach (Expr e in f[ctx.MkAdd(x, ctx.MkInt(1)), ctx.MkReal(1)].Args)
            Console.WriteLine(e);
            Console.WriteLine(f[ctx.MkAdd(x, ctx.MkInt(1)), ctx.MkReal(1)].Args[0]);
            Console.WriteLine(f[ctx.MkAdd(x, ctx.MkInt(1)), ctx.MkReal(1)].Args[0].Equals(ctx.MkAdd(x, ctx.MkInt(1))));
            Console.WriteLine(f[ctx.MkAdd(x, ctx.MkInt(1)), ctx.MkReal(1)].FuncDecl[ctx.MkInt(2), ctx.MkInt2Real((IntExpr)ctx.MkAdd(x, ctx.MkInt(1)))]);

            Console.WriteLine(ctx.MkInt(1).IsExpr);
            Console.WriteLine(ctx.MkAdd(x, ctx.MkInt(1)).IsExpr);
            Console.WriteLine(ctx.MkForall(new Expr[] { x }, ctx.MkGt(x, ctx.MkInt(0))).IsExpr);
            Console.WriteLine(ctx.MkInt(1).IsConst);
            Console.WriteLine(x.IsConst);
            Console.WriteLine(ctx.MkAdd(x, ctx.MkInt(1)).IsConst);
            Console.WriteLine(ctx.MkForall(new Expr[] { x }, ctx.MkGt(x, ctx.MkInt(0))).IsConst);

            Console.WriteLine(ctx.MkForall(new Expr[] { x }, ctx.MkGt(x, ctx.MkInt(0))).Body.Args[0]);
            Console.WriteLine(ctx.MkForall(new Expr[] { x }, ctx.MkGt(x, ctx.MkInt(0))).Body.Args[0].IsExpr);
            Console.WriteLine(ctx.MkForall(new Expr[] { x }, ctx.MkGt(x, ctx.MkInt(0))).Body.Args[0].IsConst);
            Console.WriteLine(ctx.MkForall(new Expr[] { x }, ctx.MkGt(x, ctx.MkInt(0))).Body.Args[0].IsVar);
            Console.WriteLine(x.IsVar);
            Console.WriteLine(ctx.MkITE(ctx.MkTrue(), x, ctx.MkAdd(x, ctx.MkInt(1))));

            Context ctx1 = new Context();
            Console.WriteLine(ctx1.MkITE(ctx1.MkTrue(), x.Translate(ctx1), ctx.MkAdd(x, ctx.MkInt(1)).Translate(ctx1)));
            Console.WriteLine(ctx.MkITE(ctx.MkTrue(), ctx.MkInt(1), ctx.MkInt(1)));

            Console.WriteLine(ctx.MkDistinct(x, ctx.MkAdd(x, ctx.MkInt(1)), ctx.MkAdd(x, ctx.MkInt(2))));

            Console.WriteLine(ctx1.MkAnd(ctx1.MkDistinct(x.Translate(ctx1), ctx1.MkInt(1)),
                                         ctx1.MkGt((IntExpr)x.Translate(ctx1), ctx1.MkInt(0))));
                        
        }
    }
}

File: old-regressions/python/tst5.py

# Copyright (c) 2015 Microsoft Corporation
from z3 import *

f = Function('f', IntSort(), RealSort(), IntSort())

try:
    print(f.domain(3))
except Z3Exception:
    print("Failed")

x = Int('x')

print(f(1, 1))
print(f(1, 1).sort())
print(f(x, 1).num_args())
print(f(x+1, 1).children())
print(f(x+1, 1).arg(0))
print(f(x+1, 1).arg(0).eq(x+1))
print(f(x+1, x).decl()(2, x+1))
print(is_expr(1))
print(is_expr(IntVal(1)))
print(is_expr(x + 1))
print(is_app(1))
print(is_app(IntVal(1)))
print(is_app(x + 1))
print(is_expr(ForAll(x, x > 0)))
print(is_app(ForAll(x, x > 0)))
print(is_const(IntVal(1)))
print(is_const(x))
print(is_const(x + 1))
print(is_const(ForAll(x, x > 0)))
print(ForAll(x, x > 0).body().arg(0))
print(is_expr(ForAll(x, x > 0).body().arg(0)))
print(is_app(ForAll(x, x > 0).body().arg(0)))
print(is_const(ForAll(x, x > 0).body().arg(0)))
print(is_var(ForAll(x, x > 0).body().arg(0)))
print(is_var(x))
print(If(True, x, x+1))
ctx = Context()
print(If(True, x.translate(ctx), (x+1).translate(ctx)) == If(True, 1, 1, ctx))
print(Distinct(x, x+1, x+2))
try:
    print(Distinct(1, 2, 3))
except Z3Exception:
    print("failed")
print(And(Distinct(x.translate(ctx), 1),
          x.translate(ctx) > 0))


